python/cudf/cudf/core/column/column.py

```diff
@@ -487,42 +488,67 @@ def __setitem__(self, key, value):
 
         if is_scalar(value):
             if is_categorical_dtype(self.dtype):
-                from cudf.utils.cudautils import fill_value
-
-                data = rmm.device_array(nelem, dtype=self.codes.dtype)
-                fill_value(data, self._encode(value))
-                value = build_categorical_column(
-                    categories=self.dtype.categories,
-                    codes=as_column(data),
-                    ordered=self.dtype.ordered,
-                )
-            elif value is None:
-                value = column.column_empty(nelem, self.dtype, masked=True)
-            else:
-                to_dtype = pd.api.types.pandas_dtype(self.dtype)
-                value = utils.scalar_broadcast_to(value, nelem, to_dtype)
-
-        value = column.as_column(value).astype(self.dtype)
+                value = self._encode(value)
 
-        if len(value) != nelem:
-            msg = (
-                f"Size mismatch: cannot set value "
-                f"of size {len(value)} to indexing result of size "
-                f"{nelem}"
-            )
-            raise ValueError(msg)
+            value = self.dtype.type(value) if value is not None else value
+        else:
+            if len(value) != nelem:
+                msg = (
+                    f"Size mismatch: cannot set value "
+                    f"of size {len(value)} to indexing result of size "
+                    f"{nelem}"
+                )
+                raise ValueError(msg)
+            value = column.as_column(value).astype(self.dtype)
+            if is_categorical_dtype(value.dtype):
+                value = value.cat().set_categories(self.categories)
+                assert self.dtype == value.dtype
 
-        if is_categorical_dtype(value.dtype):
-            value = value.cat().set_categories(self.categories)
-            assert self.dtype == value.dtype
+        if (
+            isinstance(key, slice)
+            and (key_stride == 1 or key_stride is None)
+            and not is_scalar(value)
+        ):
 
-        if isinstance(key, slice):
-            out = libcudf.copying.copy_range(
-                self, value, key_start, key_stop, 0
+            out = libcudfxx.copying.copy_range(
+                value, self, 0, nelem, key_start, key_stop, False
             )
+            if is_categorical_dtype(value.dtype):
+                out = build_categorical_column(
+                    categories=value.categories,
+                    codes=out,
+                    mask=out.mask,
```

Comment:
Since you're explicitly passing the offset / size:
```suggestion
                    mask=out.base_mask,
```